### PR TITLE
Jeff/step 17

### PR DESCRIPTION
**Pull request by jeff-yoo-rakuma**
# Overview
Step 17 integrates user-service to the existing task-service and its frontend app.
After the integration, a typical flow will resemble the following;
```
User login -> user-service authenticates -> user-service returns JWT -> JWT saved in client(browser) -> Client makes a request to task-service with JWT included -> task-service verifies JWT -> task-service handles request if JWT valid
```
You can create a new user on your own, or you can use the included user;
```
email: admin@example.com
password: admin
```
Please check out [README file](https://github.com/Fablic/training/blob/jeff/step_17/todos-app/README.md) for more detail and local env setup.
There is no need to review frontend implementation in detail! 🙇‍♂️ 

# Changes
## Backend
* Add JWT auth in task-service
* Add JWT auth in user-service
* Add relevant tests
## Frontend
* Add login/register pages
* Uses JWT to make requests to task-service backend

# Tests
## user-service
```
root@4d8c3709a111:/app# rspec
................................

Finished in 2.2 seconds (files took 7.54 seconds to load)
32 examples, 0 failures
```
## task-service
```
root@c95b57cc264f:/app# rspec
......................................

Finished in 2.52 seconds (files took 7.59 seconds to load)
38 examples, 0 failures
```

### CODE REVIEW

File: todos-app/backend/task_service/config/secrets.yml [hunk 1]
```diff
@@ -10,6 +10,11 @@
 # Make sure the secrets in this file are kept private
 # if you're sharing your code publicly.
 
+# Shared secrets are available across all environments.
+
+shared:
+  jwt_secret_key: <%=ENV["JWT_SECRET_KEY"] %>
```
Comment (tana2034) [inline todos-app/backend/task_service/config/secrets.yml:16]: [note]this secret key came from docker compose, and it's same to user_service one

File: todos-app/backend/task_service/app/controllers/application_controller.rb [hunk 2]
```diff
@@ -23,4 +23,30 @@ def invalid_enum(exception)
...
+      token = request.headers['Authorization'].split.pop
+      begin
+        decoded = JWT.decode(token, Rails.application.secrets.jwt_secret_key, true, algorithm: 'HS256')
+        # first element is the JWT payload, second element is the JWT header
+        decoded.first
+      rescue JWT::DecodeError
+        nil
+      end
+    end
+  end
```
Comment (tana2034) [inline todos-app/backend/task_service/app/controllers/application_controller.rb:39]: [note] JWT is included in header's Autorization with using Bearer Scheme.

File: todos-app/backend/user_service/app/controllers/application_controller.rb [hunk 3]
```diff
@@ -1,4 +1,48 @@
...
+  def invalid_enum(exception)
+    render json: { error: exception.message }, status: :unprocessable_entity
+  end
+
+  def encode_token(user, expire_after = 5 * 60)
+    payload = {
+      user_id: user.id,
+      role: user.role,
+      exp: Time.now.to_i + expire_after,
+    }
```
Comment (tana2034) [inline todos-app/backend/user_service/app/controllers/application_controller.rb:15]: [note]JWT payload includes user's information and expire date.

File: todos-app/docker-compose.yaml [hunk 4]
```diff
@@ -25,6 +25,8 @@ services:
       DB_NAME: todos-app
       DB_USER: root
       DB_PASSWORD: password
+      # the JWT key here must match the one in user-service
+      JWT_SECRET_KEY: supersecretkey
```
Comment (tana2034) [inline todos-app/docker-compose.yaml:29]: [note] it's common key

File: todos-app/frontend/src/views/Login.vue [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+    const router = useRouter()
+    const loginFailed = ref(false)
+
+    function login(event) {
+      const auth = {
+        email: event.target.email.value,
+        password: event.target.password.value
+      }
+      axios.post(process.env.VUE_APP_USER_SERVICE_BASE_URL + "/login", {auth}).then((response) => {
+        localStorage.token = response.data.token
```
Comment (tana2034) [inline todos-app/frontend/src/views/Login.vue:46]: [note] JWT stored in localStroage

**Review by tana2034 — APPROVED** [review]
I'm sorry for late reviewing 🙇 🙇 🙇 
I learned a lot from your PR. 
I looked into JWT authentication and found it suitable for microservices because it doesn't require multiple services to have information about user sessions.
(Currently, Rakuma is also looking into microservices and creating an authentication service. It's in Japanese, but if you're interested they talk about it on the [#pj_楽天メトロ ](https://fablic.slack.com/archives/C01MABHTJF3) channel)
Anyway, very informative, LGTM, and good luck with the Product training! 😄 

LGTM

**Comment by jeff-yoo-rakuma** [comment]
@tana2034 
Thank you so much for your time reviewing! 🙇

**Pull request merged**